services/rotations/rotation.service.js

```diff
@@ -0,0 +1,118 @@
+const RedisService = require("../redis");
+
+class RotationService {
+  constructor(keyName, rotationName) {
+    this.keyName = keyName;
+    this.rotationName = rotationName;
+    this.redis = RedisService.getInstance();
+  }
+
+  async #addMembers(memberList) {
+    const memberIds = memberList.map((member) => member?.id || member);
+    await this.redis.rpush(this.keyName, memberIds);
+  }
+
+  async #createNewMemberList(memberList) {
+    await this.redis.del(this.keyName);
+    await this.#addMembers(memberList);
+  }
+
+  async #getMemberList() {
+    const members = await this.redis.lrange(this.keyName, 0, -1);
+    return members;
+  }
+
+  async #swapMembers(members) {
+    const [firstMember, secondMember] = members.map((member) => member.id);
+    const currentList = await this.#getMemberList();
+
+    const firstMemberIndex = currentList.indexOf(firstMember);
+    const secondMemberIndex = currentList.indexOf(secondMember);
+
+    currentList[firstMemberIndex] = secondMember;
+    currentList[secondMemberIndex] = firstMember;
+
+    await this.#createNewMemberList(currentList);
+  }
+
+  async #getDisplayNames(members, server) {
+    this.displayNames = members.map(async (memberId) => {
+      const member = await server.members.fetch(memberId);
+      if (member.nickname) {
+        return member.nickname;
+      }
+      return member.user.username;
+    });
+
+    return Promise.all(this.displayNames);
+  }
+
+  async #getFormattedMemberList(server) {
+    const members = await this.#getMemberList();
+    const membersDisplayNames = await this.#getDisplayNames(members, server);
+    const formattedMemberList = membersDisplayNames.reduce(
+      (acc, displayname) => `${acc} ${displayname} >`,
+      ""
+    );
+    if (formattedMemberList) {
+      return formattedMemberList;
+    }
+    return "No members";
+  }
+
+  async #rotateMemberList(interaction) {
+    const memberToPing = await this.redis.lpop(this.keyName)
+    await this.#addMembers([memberToPing])
+
+    const formattedMembers = await this.#getFormattedMemberList(
+      interaction.guild
+    );
+    const reply = `<@${memberToPing}> it's your turn for the ${this.rotationName} rotation.\nThe ${this.rotationName} rotation order is now ${formattedMembers}.`;
+    interaction.reply(reply);
+  }
+
+  async #removeMember(member) {
+    const memberId = member.id;
+    await this.redis.lrem(this.keyName, 0, memberId);
+  }
+
+  #getMembers(interactionOptions) {
+    this.members = [];
+    for (let i = 0; i < 10; i += 1) {
+      this.members.push(interactionOptions.getUser(`user${i}`));
+    }
+    return this.members.filter((member) => !!member);
+  }
+
+  async handleInteraction(interaction) {
+    const actionType = interaction.options.getSubcommand();
+
+    const members = this.#getMembers(interaction.options);
+
+    let replyModifier = "updated to";
+
+    switch (actionType) {
+      case "add":
+        await this.#addMembers(members);
```

Comment:
One small nit: can we make sure that we cannot add members that are already in the list?

Reply:
I'm not sure how needed that really is since this is just meant to be an internal tool, but it shouldn't add _much_ more complexity so I suppose I'm not against it.